sharedbuffers/mapped_struct.py

```diff
@@ -2624,18 +2624,44 @@ def _merge_all(parts, dtype):
         return _merge_all(nparts, dtype)
 
 @cython.cfunc
-@cython.locals(parts = list, discard_duplicate_keys = cython.bint, discard_duplicates = cython.bint)
-def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplicates):
+@cython.locals(discard_duplicate_keys = cython.bint, discard_duplicates = cython.bint, copy = cython.bint,
+    nout = cython.size_t, nmin = cython.size_t, out_start = cython.size_t)
+def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplicates, copy):
     if discard_duplicate_keys or discard_duplicates:
         # What numpy.unique does, but using MUCH less RAM
         vpart = apart.view(struct_dt)
         vpart.sort(0)
         if discard_duplicate_keys:
-            flags = numpy.concatenate([[True], apart[1:,0] != apart[:-1,0]])
+            flags = apart[1:,0] != apart[:-1,0]
         elif discard_duplicates:
-            flags = numpy.concatenate([[True], vpart[1:,0] != vpart[:-1,0]])
-        if not numpy.all(flags):
-            apart = apart[flags]
+            flags = vpart[1:,0] != vpart[:-1,0]
+        np = numpy
+        if not np.all(flags):
+            if copy:
+                # Simple boolean indexing, best implementation for constructing a deduplicated copy
+                flags = np.concatenate([[True], flags])
+                apart = apart[flags]
+            else:
+                # In-place compress operation, best option for non-copying deduplication, to avoid
+                # having to allocate temporary workspace proportional to output size. This
+                # way only uses a fixed amount of temporary workspace and builds a compressed
+                # result into the input array directly.
+                nmin = np.argmin(flags)
+                count_nonzero = np.count_nonzero
+                start = nmin
```

Comment:
Why can't you do `start = nmin + 1` to prevent doing the `+1` all the time?

Reply:
If you read carefully, you'll see that the start/end/chunk_end variables get used in both contexts, those that need the +1 and those that don't.